src/uct/tcp/tcp_sockcm_ep.c

```diff
@@ -80,33 +93,90 @@ ucs_status_t uct_tcp_sockcm_ep_progress_send(uct_tcp_sockcm_ep_t *cep)
+                      cep, (cep->state & UCT_TCP_SOCKCM_EP_ON_SERVER) ? "server" : "client",
+                      cep->comm_ctx.length, cep->comm_ctx.offset);
         }
-        return status;
+
+        goto out;
     }
 
     cep->comm_ctx.offset += sent_length;
     ucs_assert(cep->comm_ctx.offset <= cep->comm_ctx.length);
-    cep->state           |= UCT_TCP_SOCKCM_EP_SENDING;
 
     if (uct_tcp_sockcm_ep_is_tx_rx_done(cep)) {
+        ucs_assert(status == UCS_OK);
         cep->state |= UCT_TCP_SOCKCM_EP_DATA_SENT;
-        uct_tcp_sockcm_ep_init_comm_ctx(cep);
 
-        /* wait for a reply from the peer */
-        status = ucs_async_modify_handler(cep->fd, UCS_EVENT_SET_EVREAD);
-        if (status != UCS_OK) {
-            ucs_error("failed to modify %d event handler to "
-                      "UCS_EVENT_SET_EVREAD: %s", cep->fd,
-                      ucs_status_string(status));
+        /* on the client side - if completed sending a message after the notify
+         * call was invoked, then this message is the notify message */
+        if (cep->state & UCT_TCP_SOCKCM_EP_CLIENT_NOTIFY_CALLED) {
+            ucs_assert(cep->state & UCT_TCP_SOCKCM_EP_ON_CLIENT);
+            cep->state |= UCT_TCP_SOCKCM_EP_CLIENT_NOTIFY_SENT;
         }
+
+        uct_tcp_sockcm_ep_init_comm_ctx(cep);
+
+        /* wait for a message from the peer */
+        events = UCS_EVENT_SET_EVREAD;
+    } else {
+        /* continue the sending when possible, and handle potential disconnect */
+        events = UCS_EVENT_SET_EVREAD | UCS_EVENT_SET_EVWRITE;
     }
 
-    return UCS_OK;
+    status = ucs_async_modify_handler(cep->fd, events);
+    if (status != UCS_OK) {
+        ucs_error("failed to modify %d event handler to %d: %s",
+                  cep->fd, events, ucs_status_string(status));
+    }
+
+out:
+    return status;
+}
+
+static const char *uct_tcp_sockcm_cm_ep_peer_addr_str(uct_tcp_sockcm_ep_t *cep,
+                                                      char *buf, size_t max)
+{
+    struct sockaddr_storage remote_dev_addr = {0}; /* Suppress Clang false-positive */
+    socklen_t remote_dev_addr_len;
+    ucs_status_t status;
+
+    /* get the device address of the remote peer associated with the connected fd */
+    status = ucs_socket_getpeername(cep->fd, &remote_dev_addr, &remote_dev_addr_len);
+    if (status != UCS_OK) {
+        return NULL;
+    }
+
+    return ucs_sockaddr_str((const struct sockaddr*)&remote_dev_addr, buf, max);
+}
+
+ucs_status_t uct_tcp_sockcm_cm_ep_conn_notify(uct_ep_h ep)
+{
+    uct_tcp_sockcm_ep_t *cep                = ucs_derived_of(ep, uct_tcp_sockcm_ep_t);
+    uct_tcp_sockcm_t *tcp_sockcm            = uct_tcp_sockcm_ep_get_cm(cep);
+    uct_tcp_sockcm_priv_data_hdr_t *hdr;
+    char peer_str[UCS_SOCKADDR_STRING_LEN];
+    ucs_status_t status;
+
+    UCS_ASYNC_BLOCK(tcp_sockcm->super.iface.worker->async);
+
+    hdr = (uct_tcp_sockcm_priv_data_hdr_t*)cep->comm_ctx.buf;
```

Comment:
maybe do some assertion that we are not sending anything else already? e.g that cep->comm_ctx.buf is available for reuse

Reply:
how can we tell that it can be reused? cep->comm_ctx is used both for sending and receving.
maybe check that the DATA_SENT flag is on?

Reply:
> maybe check that the DATA_SENT flag is on?

Yes